identity/background.js

```diff
@@ -0,0 +1,237 @@
+             '&response_type=token' +
+             '&redirect_uri=' + encodeURIComponent(redirectURL) +
+             '&scope=' + encodeURIComponent(this.scopes.join(' '));
+    },
+
+    verifyToken(token) {
+      return new Promise((resolve, reject) => {
+        let url = 'https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=' + token;
+        let xhr = new XMLHttpRequest();
+        xhr.open('GET', url);
+        xhr.onload = (e) => {
+          let r = e.target;
+          if (r.status === 200) {
+            resolve(JSON.parse(r.responseText));
+          } else {
+            reject([r.status, r.responseText]);
+          }
+        };
+        xhr.onerror = (e) => {
+          let r = e.target;
+          let response = JSON.parse('"'+r.responseText+'"');
+          reject([r.status, response]);
+        };
+        xhr.send(null);
+      });
+    },
+
+    authResult(params, redirectUri) {
+      return new Promise((resolve, reject) => {
+        if (params.get("error")) {
+          reject(new Error(params.get("error_description")));
+        } else if (params.get("id_token")) {
+          this.token = params.get("id_token");
+          resolve(this.token);
+        } else if (params.get("access_token")) {
+          let token = params.get("access_token");
+          this.verifyToken(token).then(verified => {
+            // TODO verify the data
+            // https://developers.google.com/identity/protocols/OAuth2UserAgent
+            this.verified = verified;
+            this.token = token;
+            resolve(this.token);
+          });
+        } else
+          reject(new Error("Invalid response, no code or token"));
+      });
+    },
+
+    getUserInfo(interactive) {
+      return new Promise((resolve, reject) => {
+        if (this.user_info) {
+          resolve(this.user_info);
+          return;
+        } else
+        if (!interactive) {
+          reject(new Error("user not logged in"));
+          return;
+        }
+        xhrWithAuth(this, 'GET', 'https://www.googleapis.com/oauth2/v1/userinfo?alt=json', interactive).then(response => {
+          this.user_info = JSON.parse(response.response);
+          resolve(this.user_info);
+        });
+      });
+    }
+  },
+};
+
+function parseSearchParams(redirectUri) {
+  let m = redirectUri.match(/[#\?](.*)/);
+  if (!m || m.length < 1)
+    return {};
+  return new URLSearchParams(m[1].split("#")[0]);
+}
+
+function authorize(provider, interactive) {
+  let redirectUri = chrome.identity.getRedirectURL('/callback');
+
+  return new Promise((resolve, reject) => {
+    if (provider.token) {
+      resolve(provider.token);
+      return;
+    }
+
+    let options = {
+      interactive: interactive,
+      url: provider.getAuthURL(redirectUri)
+    };
+
+    browser.identity.launchWebAuthFlow(options, function(redirectURL) {
+      if (chrome.runtime.lastError) {
+        reject(new Error(chrome.runtime.lastError.message));
```

Comment:
> We've switched all the examples to browser now.

Wait, really? That means the examples won't run in Chrome though, right? That seems excessively harsh, especially if we want to encourage devs to write cross-platform extensions.